decidim/decidim
# Improve components and spaces sharing with administrable tokens

#### :tophat: What? Why?

This PR improves the current behaviour of the share tokens introduced in #6271 with a full CRUD operational mode.

New features allow to:
1. Create expirable tokens
2. Create tokens only for logged user
3. Create custom tokens (instead of random)
4. Allows to share the whole unpublished participatory space instead of component by component
5. Remembers the share token in session so it is possible to navigate through the different pages 

This improvement is funded by the city of Zurich.

#### :pushpin: Related Issues
*Link your PR to an issue*
- Related to ##6271
- Fixes https://github.com/openpoke/decidim/issues/76

#### Testing

case 1:
- As an admin, go to a participatory space, edit it. Set it to the "unpublished" state.
- Go to the "share tokens" menu
- Create/edit/destroy tokens at will
- Copy the url for a created token
- Visit the public page of participatory space using the copied link with another browser/session
- See that you can navigate through the process even that's unpublished.

Case 2:
- As an admin, go to the "components" page of a participatory space.
- Click on the "share" icon
- Manage tokens
- Use some of the urls with token to visit an the "unpublished" component.
- See that you can navigate through the component even that's unpublished.

### :camera: Screenshots

![image](https://github.com/user-attachments/assets/364cfc94-c4d7-4bbb-a603-9967f769b5d8)
![image](https://github.com/user-attachments/assets/c14f7550-768d-4ab5-a291-18cf250b9e03)
![image](https://github.com/user-attachments/assets/9f16fd55-b2a2-4b9f-9161-ff574ecd0493)
![image](https://github.com/user-attachments/assets/ee7c9a0d-8660-4e7c-a50e-f5b9055c1578)

:hearts: Thank you!
